pkg/internal/ebpf/common/kafka_detect_transform.go

```diff
@@ -182,31 +183,64 @@ func isValidClientID(buffer []byte, realClientIDSize int) bool {
 	return isValidKafkaString(buffer, len(buffer), realClientIDSize, true)
 }
 
-func getTopicName(pkt []byte, offset int) (string, error) {
+func getTopicName(pkt []byte, offset int, op Operation, apiVersion int16) (string, error) {
+	if apiVersion >= 13 { // topic name is only a UUID, no need to parse it
+		return "", nil
+	}
+
 	offset += 4
 	if offset > len(pkt) {
 		return "", errors.New("invalid buffer length")
 	}
-	topicNameSize := int16(binary.BigEndian.Uint16(pkt[offset:]))
-	if topicNameSize <= 0 || topicNameSize > 255 {
-		return "", errors.New("invalid topic name size")
+	topicNameSize, err := getTopicNameSize(pkt, offset, op, apiVersion)
+	if err != nil {
+		return "", err
 	}
 	offset += 2
 
 	if offset > len(pkt) {
 		return "", nil
 	}
-	maxLen := offset + int(topicNameSize)
+	maxLen := offset + topicNameSize
 	if len(pkt) < maxLen {
 		maxLen = len(pkt)
 	}
 	topicName := pkt[offset:maxLen]
+	if op == Fetch && apiVersion > 11 { // topic name has the following format: uuid\x00\x02\tTOPIC\x02\x00
+		topicNames := strings.Split(string(topicName), "\t")
+		if len(topicNames) >= 2 {
+			topicNames := strings.Split(string(topicNames[1]), "\x02\x00")
+			if len(topicNames) > 0 {
+				topicName = []byte(topicNames[0])
+			}
+		} else {
+			topicName = []byte(topicNames[0])
+		}
+	}
 	if isValidKafkaString(topicName, len(topicName), int(topicNameSize), false) {
 		return string(topicName), nil
 	}
 	return "", errors.New("invalid topic name")
 }
 
+func getTopicNameSize(pkt []byte, offset int, op Operation, apiVersion int16) (int, error) {
+	topicNameSize := 0
+	if (op == Produce && apiVersion > 7) || (op == Fetch && apiVersion > 11) { // topic is a compact string
+		var err error
+		topicNameSize, err = readUnsignedVarint(pkt[offset+1:])
+		topicNameSize--
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		topicNameSize = int(binary.BigEndian.Uint16(pkt[offset:]))
+	}
+	if topicNameSize <= 0 || topicNameSize > 255 {
```

Comment:
Is the topicNameSize > 255 still a valid check if we are > 11? Variable size int encoding makes me think it can now be larger than 255...

Reply:
removed the upper limit